File: showcases/arqitect/example-tool/tool.py
"""
currency-converter — Arqitect MCP tool that converts between currencies.

Reads newline-delimited JSON-RPC 2.0 requests from stdin and writes responses
to stdout. The armor manifest (armor.json) in this directory declares that this
tool may only reach api.frankfurter.app:443. All other network access,
filesystem writes, and subprocess spawning are denied by the MCP Armor broker.

Supported methods:
    tools/list  — returns the list of available tools
    tools/call  — dispatches to the convert_currency tool handler

Tools:
    convert_currency — converts an amount between two ISO 4217 currency codes
"""

from __future__ import annotations

import json
import sys
import urllib.error
import urllib.request
from typing import Any

FRANKFURTER_BASE_URL = "https://api.frankfurter.app/latest"


def fetch_exchange_rate(from_currency: str, to_currency: str) -> float:
    """
    Fetch the live exchange rate between two currencies from Frankfurter.

    Args:
        from_currency: ISO 4217 source currency code (e.g. "USD").
        to_currency: ISO 4217 target currency code (e.g. "EUR").

    Returns:
        The exchange rate as a float (amount of to_currency per 1 from_currency).

    Raises:
        ValueError: If the currency pair is not supported.
        OSError: If the API request fails.
    """
    url = f"{FRANKFURTER_BASE_URL}?from={from_currency}&to={to_currency}"
    try:
        with urllib.request.urlopen(url, timeout=10) as response:
            data = json.loads(response.read())
    except urllib.error.HTTPError as err:
        raise ValueError(f"Currency API error {err.code}: {err.reason}") from err

    rates = data.get("rates", {})
    if to_currency not in rates:
        raise ValueError(f"Unsupported currency pair: {from_currency} -> {to_currency}")
    return float(rates[to_currency])


def convert_currency(amount: float, from_currency: str, to_currency: str) -> dict[str, Any]:
    """
    Convert an amount from one currency to another using live exchange rates.

    Args:
        amount: Amount to convert (must be positive).
        from_currency: ISO 4217 source currency code.
        to_currency: ISO 4217 target currency code.

    Returns:
        MCP tools/call result dict with a text content item.

    Raises:
        ValueError: If the amount is not positive or the currencies are invalid.
    """
    if amount <= 0:
        raise ValueError("Amount must be positive")
    if not from_currency.strip() or not to_currency.strip():
        raise ValueError("Currency codes must not be empty")

    rate = fetch_exchange_rate(from_currency.upper(), to_currency.upper())
    converted = round(amount * rate, 2)

    return {
        "content": [
            {
                "type": "text",
                "text": (
                    f"{amount} {from_currency.upper()} = {converted} {to_currency.upper()} "
                    f"(rate: 1 {from_currency.upper()} = {rate} {to_currency.upper()})"
                ),
            }
        ]
    }


def handle_tools_list() -> dict[str, Any]:
    """
    Return the MCP tools/list result for this tool.

    Returns:
        Dict with a 'tools' key listing all available tools and their schemas.
    """
    return {
        "tools": [
            {
                "name": "convert_currency",
                "description": "Convert an amount from one currency to another using live exchange rates.",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "amount": {"type": "number", "description": "Amount to convert"},
                        "from": {"type": "string", "description": "ISO 4217 source currency code (e.g. USD)"},
                        "to": {"type": "string", "description": "ISO 4217 target currency code (e.g. EUR)"},
                    },
                    "required": ["amount", "from", "to"],
                },
            }
        ]
    }


def handle_tools_call(params: dict[str, Any]) -> dict[str, Any]:
    """
    Dispatch a tools/call request to the appropriate tool handler.

    Args:
        params: JSON-RPC params dict with 'name' and 'arguments' keys.

    Returns:
        MCP tools/call result dict.

    Raises:
        ValueError: If the tool name is unknown or arguments are invalid.
    """
    tool_name = params.get("name")
    arguments = params.get("arguments", {})

    if tool_name != "convert_currency":
        raise ValueError(f"Unknown tool: {tool_name}")

    amount = arguments.get("amount")
    from_currency = arguments.get("from")
    to_currency = arguments.get("to")

    if amount is None or from_currency is None or to_currency is None:
        raise ValueError("convert_currency requires 'amount', 'from', and 'to' arguments")

    return convert_currency(float(amount), str(from_currency), str(to_currency))


def process_request(request: dict[str, Any]) -> dict[str, Any]:
    """
    Process a single parsed JSON-RPC 2.0 request and return a response dict.

    Args:
        request: Parsed JSON-RPC request object.

    Returns:
        JSON-RPC 2.0 response dict (success or error).
    """
    request_id = request.get("id")
    method = request.get("method")
    params = request.get("params", {})

    try:
        if method == "tools/list":
            result = handle_tools_list()
        elif method == "tools/call":
            result = handle_tools_call(params)
        else:
            return {"jsonrpc": "2.0", "id": request_id, "error": {"code": -32601, "message": f"Method not found: {method}"}}
        return {"jsonrpc": "2.0", "id": request_id, "result": result}
    except ValueError as err:
        return {"jsonrpc": "2.0", "id": request_id, "error": {"code": -32000, "message": str(err)}}
    except OSError as err:
        return {"jsonrpc": "2.0", "id": request_id, "error": {"code": -32000, "message": f"Network error: {err}"}}


def main() -> None:
    """Read newline-delimited JSON-RPC requests from stdin and write responses to stdout."""
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            request = json.loads(line)
        except json.JSONDecodeError:
            response = {"jsonrpc": "2.0", "id": None, "error": {"code": -32700, "message": "Parse error"}}
            sys.stdout.write(json.dumps(response) + "\n")
            sys.stdout.flush()
            continue

        response = process_request(request)
        sys.stdout.write(json.dumps(response) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()

File: showcases/openclaw/example-skill/tool.js
'use strict';

/**
 * weather — OpenClaw skill that fetches current weather for a city.
 *
 * MCP tool interface: reads JSON-RPC requests from stdin, writes responses
 * to stdout. The armor manifest (armor.json) in this directory declares that
 * this skill may only reach api.openweathermap.org:443 and api.open-meteo.com:443.
 * All other network access, filesystem writes, and subprocess spawning are denied.
 *
 * Supported methods:
 *   - tools/list      — returns the list of available tools
 *   - tools/call      — dispatches to the appropriate tool handler
 *
 * Tools:
 *   - get_weather     — returns current weather for a given city name
 */

const https = require('https');
const readline = require('readline');

const OPENWEATHER_API_KEY = process.env.OPENWEATHER_API_KEY ?? '';
const OPENWEATHER_BASE_URL = 'https://api.openweathermap.org/data/2.5/weather';

/**
 * Fetches current weather data for a city from the OpenWeatherMap API.
 *
 * @param {string} city - City name to look up
 * @returns {Promise<{ temperature: number, description: string, humidity: number }>}
 * @throws {Error} If the API request fails or the city is not found
 */
function fetchWeather(city) {
  return new Promise((resolve, reject) => {
    const url = `${OPENWEATHER_BASE_URL}?q=${encodeURIComponent(city)}&appid=${OPENWEATHER_API_KEY}&units=metric`;
    https.get(url, (res) => {
      let body = '';
      res.on('data', (chunk) => { body += chunk; });
      res.on('end', () => {
        if (res.statusCode !== 200) {
          reject(new Error(`Weather API error: ${res.statusCode} ${body}`));
          return;
        }
        const data = JSON.parse(body);
        resolve({
          temperature: data.main.temp,
          description: data.weather[0].description,
          humidity: data.main.humidity,
        });
      });
    }).on('error', reject);
  });
}

/**
 * Handles a tools/list request.
 *
 * @returns {object} MCP tools/list result
 */
function handleToolsList() {
  return {
    tools: [
      {
        name: 'get_weather',
        description: 'Get the current weather for a city.',
        inputSchema: {
          type: 'object',
          properties: {
            city: { type: 'string', description: 'City name (e.g. "London")' },
          },
          required: ['city'],
        },
      },
    ],
  };
}

/**
 * Handles a tools/call request by dispatching to the named tool.
 *
 * @param {string} toolName - Name of the tool to invoke
 * @param {Record<string, unknown>} toolArgs - Arguments for the tool
 * @returns {Promise<object>} MCP tools/call result
 */
async function handleToolCall(toolName, toolArgs) {
  if (toolName !== 'get_weather') {
    throw new Error(`Unknown tool: ${toolName}`);
  }
  if (typeof toolArgs.city !== 'string' || toolArgs.city.trim() === '') {
    throw new Error('get_weather requires a non-empty city argument');
  }
  const weather = await fetchWeather(toolArgs.city.trim());
  return {
    content: [
      {
        type: 'text',
        text: `Weather in ${toolArgs.city}: ${weather.temperature}°C, ${weather.description}, humidity ${weather.humidity}%.`,
      },
    ],
  };
}

/**
 * Builds a JSON-RPC 2.0 success response.
 *
 * @param {number | string} id
 * @param {unknown} result
 * @returns {string} Serialized JSON-RPC response
 */
function successResponse(id, result) {
  return JSON.stringify({ jsonrpc: '2.0', id, result });
}

/**
 * Builds a JSON-RPC 2.0 error response.
 *
 * @param {number | string | null} id
 * @param {number} code
 * @param {string} message
 * @returns {string} Serialized JSON-RPC response
 */
function errorResponse(id, code, message) {
  return JSON.stringify({ jsonrpc: '2.0', id, error: { code, message } });
}

/**
 * Processes a single parsed JSON-RPC request and returns the response string.
 *
 * @param {object} request - Parsed JSON-RPC request object
 * @returns {Promise<string>} Serialized JSON-RPC response
 */
async function processRequest(request) {
  const { id, method, params } = request;
  try {
    if (method === 'tools/list') {
      return successResponse(id, handleToolsList());
    }
    if (method === 'tools/call') {
      const result = await handleToolCall(params.name, params.arguments ?? {});
      return successResponse(id, result);
    }
    return errorResponse(id, -32601, `Method not found: ${method}`);
  } catch (err) {
    return errorResponse(id, -32000, err.message);
  }
}

/** Reads newline-delimited JSON-RPC requests from stdin and writes responses to stdout. */
async function main() {
  const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
  for await (const line of rl) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    let request;
    try {
      request = JSON.parse(trimmed);
    } catch {
      process.stdout.write(errorResponse(null, -32700, 'Parse error') + '\n');
      continue;
    }
    const response = await processRequest(request);
    process.stdout.write(response + '\n');
  }
}

main().catch((err) => {
  process.stderr.write(`Fatal error: ${err.message}\n`);
  process.exit(1);
});
